ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -530,11 +552,45 @@ public boolean addBlobStore(ReplicaId replica) {
     logger.info("New store is successfully added into StorageManager");
     return true;
   }
+
+  /**
+   * Build inmemory state for file copy based replication post filecopy is completed.
+   * @param replica the {@link ReplicaId} of the {@link Store} for which store needs to be built
+   */
   @Override
-  public void buildStateForFileCopy(String partitionName){
-    // no-op
+  public void buildStateForFileCopy(ReplicaId replica){
+    if (replica == null) {
+      logger.error("ReplicaId is null");
+      throw new StateTransitionException("ReplicaId null is not found in clustermap for " + currentNode, ReplicaNotFound);
+    }
+    PartitionId partitionId = replica.getPartitionId();
+
+    if (!addBlobStoreForFileCopy(replica)){
+      // We have decreased the available disk space in HelixClusterManager#getDiskForBootstrapReplica. Increase it
+      // back since addition of store failed.
+      replica.getDiskId().increaseAvailableSpaceInBytes(replica.getCapacityInBytes());
```

Comment:
We should have cleanup before decreasing the space on disk.

Reply:
Or, we should just throw exception and let Helix take care of exception and wait for downward transition.

Reply:
The disk space is already reduced in pre-filecopy steps since diskAllocator reserves the space. This increase in space is to undo that reservation done earlier.

Reply:
1. Shouldn't we reduce the space on disk after removing the directory based on success or failure of tryRemoveFailedBootstrapBlobStore().
2. Update state in Helix as well.
Or
We can delegate the whole responsibility to the downward state transition of clean up in Helix.

Reply:
1. Before FIleCopy, we reduce the space available. Since we fail out, we have to increase the space now.